enhancements/monitoring/alert-overrides.md

```diff
@@ -0,0 +1,331 @@
+by operators and cannot be modified, and the addition of custom rules is
+explicitly unsupported at the moment.  This proposal outlines a solution for
+allowing additional user-defined rules and limited overrides of existing
+platform alerting rules.
+
+## Motivation
+
+OpenShift is an opinionated platform, and the default alerting rules have of
+course been carefully crafted to match what we think the majority of users need.
+Nevertheless, users have repeatedly asked for the ability to modify alerting
+rules to adjust thresholds, or to simply add labels and annotations to aide in
+routing alert notifications, as well as a supported method of adding additional
+alerting rules.
+
+Users currently have the ability to add additional rules from a technical
+standpoint -- by simply creating new `PrometheusRule` objects -- but this is
+[explicitly unsupported][1] at the moment.  The reason being that these
+resources are handled directly by the prometheus-operator, making it difficult
+to perform OpenShift specific defaulting and validation.  A concrete example
+being that there is no way to prevent users from creating new recording rules
+that may overwhelm the platform Prometheus instance.
+
+We've also encountered bugs and unexpected interactions that cause alerts to
+fire erroneously.  A supported method of patching platform alerting rules
+provides a more flexible method of applying temporary fixes in such cases.
+
+The goal of this proposal is to enable the addition of custom alerting rules,
+and limited ability to override existing platform alerting rules via a resource
+under the direct control of the cluster-monitoring-operator.
+
+### Goals
+
+- Give users a single, supported, and documented place to add additional rules,
+  in a way that allows the cluster-monitoring-operator to provide its own
+  defaulting, validation, etc.
+- Allow users to apply overrides to alerting rules shipped by the platform.
+- Allow users to drop unwanted platform alerts.
+- User-supplied overrides are merged with platform rules -- updates to fields
+  not modified by users continue to get updates.
+- Resources shipped by operators remain unmodified.  No action needed from
+  OpenShift developers.
+
+### Non-Goals
+
+- Only alerting rules are available for modification.  Not recording rules or
+  other monitoring configuration.
+
+## Proposal
+
+The cluster-monitoring-operator will be extended with a new controller that
+watches an instance of a new resource type with the following structure:
+
+```yaml
+apiVersion: v1alpha1
+kind: alertoverrides.monitoring.openshift.io
+metadata:
+  name: alert-overrides
+  namespace: openshift-monitoring
+spec:
+  # List of overrides for alerts in platform namespaces.
+  overrides:
+  - selector: # required
+      alert: "KubeAPIErrorBudgetBurn" # required
+      matchLabels: # optional, but multiple matches is an error reported in status.
+        severity: "critical"
+        long: "1h"
+    action: "patch"
```

Comment:
Instead of having the action as a value, have we considered having it as a key? This would allow us to define different parameters for different actions in the future if such a need arises. E.g.
```
spec:
  patch:
    ...
  drop:
    ...
```

Reply:
I don't have a super strong opinion on this. The way it is now somehow feels like a better middle ground between being Kubernetes-ish, and looking a bit like the existing Prometheus / Alertmanager configs to me. But if the consensus is to have a separate key for each action, that works too.

Good point on changing the parameters. Though, I guess we could already add new optional parameters in the existing version, and anything else probably needs a bump to the API resource version anyway, right?

Reply:
I think the advantage of separate keys is that we can have different configuration options for each operation. This makes it more visible to the user which option can be used for each action, and also adds some "typing" to action parameters.

But anyway, it's just an option to consider. I don't think it's really a show stopper to have the action as a value.

Reply:
Agree. This is easier to read when we divide the list into patch, drop, add sections.